src/java/org/apache/cassandra/streaming/StreamSession.java

```diff
@@ -1027,9 +1022,30 @@ public void receive(IncomingStreamMessage message)
     public void progress(String filename, ProgressInfo.Direction direction, long bytes, long total)
     {
         ProgressInfo progress = new ProgressInfo(peer, index, filename, direction, bytes, total);
+        updateMetricsOnProgress(progress);
         streamResult.handleProgress(progress);
     }
 
+    private void updateMetricsOnProgress(ProgressInfo progress)
+    {
+        ProgressInfo.Direction direction = progress.direction;
+        long lastSeenBytesStreamedForProgress = lastSeenBytesStreamed.getOrDefault(progress, 0L);
+        long newBytesStreamed = progress.currentBytes - lastSeenBytesStreamedForProgress;
+        if (direction == ProgressInfo.Direction.OUT)
+        {
+            StreamingMetrics.totalOutgoingBytes.inc(newBytesStreamed);
+            metrics.outgoingBytes.inc(newBytesStreamed);
+        }
+
+        else if (direction == ProgressInfo.Direction.IN)
```

Comment:
ProgressInfo.Direction can be only IN or OUT. If it is not OUT, it is IN, so we do not need `else if`, is not `else` good enough? Or direction can be null?

Reply:
In theory `direction` can be null, but in practice, it doesn't appear to be (all calls to `StreamSession::progress` seem to set a value).

Instead of adding the assert, what about using the `@Nonnull` annotation to prevent anyone from passing null in the future. Then we can just use `else`.